### PR TITLE
Teams automated functional test plan 

### LINKED ISSUE #570: Implement initial Teams Functional Test Plan
# 💡 Summary #

Create test plan for teams to support automated functional testing.  

## Motivation and context ##

Supports automated testing of products

This would be useful because it reduces human touch labor. 

## Implementation notes ##

Please provide details for implementation, such as:

- Create yanl test plan

## Acceptance criteria ##

How do we know when this work is done?

- [ ] Test plan executes without any undocumented errors

### PR DESCRIPTION
**Pull request by crutchfield**
#  #

## 🗣 Description ##

Created a test plan for Teams product.  The plan has a variant for GCC/DoD tenants.  
The 55 test cases exercise signifigant portions of the code with end to end tests. 

Closes #570 

## 💭 Motivation and context ##

Automated test plans are needed to document testing, reduce human touch labor and errors. 

## 🧪 Testing ##

Run Teams test plan as a user. 

## ✅ Pre-approval checklist ##

- [X] This PR has an informative and human-readable title.
- [X] Changes are limited to a single goal - *eschew scope creep!*
- [X] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [X] All relevant type-of-change labels have been added.
- [X] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [X] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [X] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [X] Tests have been added and/or modified to cover the changes in this PR.
- [X] All new and existing tests pass as expected.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Create a release.

NOTE: MS.TEAMS.1.5v1 fails. SHould be corrected with #452, #506, #554

### CODE REVIEW

**Comment by crutchfield** [comment]
Run PowerShell fails due to One Drive #578, #580
Run Opa Test fails due to MS.Teams.1.5v1 (see note above)

**Review by kennethpalmer — CHANGES_REQUESTED** [review]
teams.testplan.yaml is failing on all MS.TEAMS.1.5 v1 tests because there are multiple tests per policy ID

**Comment by kennethpalmer** [comment]
Cannot excercise gcc variant as I dont have permissions

**Comment by crutchfield** [comment]
> teams.testplan.yaml is failing on all MS.TEAMS.1.5 v1 tests because there are multiple tests per policy ID

There is a note above identifying the issues that cover this.  1.5 was not expected to pass due to issue. 

NOTE: MS.TEAMS.1.5v1 fails. SHould be corrected with https://github.com/cisagov/ScubaGear/issues/452, https://github.com/cisagov/ScubaGear/issues/506, https://github.com/cisagov/ScubaGear/issues/554

File: Testing/Functional/Products/TestPlans/teams.testplan.yaml [hunk 1]
```diff
@@ -62,10 +50,10 @@ TestPlan:
           - Command: Set-CsTeamsMeetingPolicy
             Splat:
               AutoAdmittedUsers: EveryoneInCompany
-            Identity: Global
+              Identity: Global
         Postconditions: []
         ExpectedResult: false
-      - TestDescription: Dialin not bypass lobbyEveryone not auto admitted user
+      - TestDescription: MS.TEAMS.1.3v1 Compliant - Dialin not bypass lobbyEveryone not auto admitted user
```
Comment (nanda-katikaneni) [inline Testing/Functional/Products/TestPlans/teams.testplan.yaml:56]: Misspell in the TestDescription (add ; after lobby)

File: Testing/Functional/Products/TestPlans/teams.testplan.yaml [hunk 2]
```diff
@@ -77,315 +65,367 @@ TestPlan:
...
+          - Command: 'Get-CsTeamsMeetingPolicy | % {Set-CsTeamsMeetingPolicy -Identity $($_.Identity) -AllowPSTNUsersToBypassLobby $true }'
+        Postconditions: []
+        ExpectedResult: false   
+      - TestDescription: MS.TEAMS.1.5v1 Non-compliant - AutoAdmittedUsers = Everyone
+        Preconditions:
+          - Command: 'Get-CsTeamsMeetingPolicy | % {Set-CsTeamsMeetingPolicy -Identity $($_.Identity) -AutoAdmittedUsers Everyone }'
+          - Command: 'Get-CsTeamsMeetingPolicy | % {Set-CsTeamsMeetingPolicy -Identity $($_.Identity) -AllowPSTNUsersToBypassLobby $true }'
+        Postconditions: []
+        ExpectedResult: false        
+      - TestDescription: MS.TEAMS.1.5v1 Non-compliant - AutoAdmittedUsers = OrganizerOnly
```
Comment (nanda-katikaneni) [inline Testing/Functional/Products/TestPlans/teams.testplan.yaml:134]: Will revisit policy 1.5 test after its planned change but note that there are a couple combinations of AutoAdmittedUsers and AllowPSTNUsersToBypassLobby are not allowed (overriden) and can be deleted from the combination of tests to be performed. For example Non-Complaint test where AutoAdmittedUsers is OrganizerOnly - will auto set AllowPSTNUsersToBypaasLobby to false (no option to make it true). Similarly the combo of 'Everyone' and 'false' is not allowed. Think those two combinations can be removed from the test cases.

**Review by nanda-katikaneni — COMMENTED** [review]
Looks good except for couple minor comments (one is a spellcheck and another is about the pending 1.5). Will revisit the 1.5 after the changes are made.

File: Testing/Functional/Products/TestPlans/teams.testplan.yaml (hunk 2 above)
Comment (kennethpalmer) [reply Testing/Functional/Products/TestPlans/teams.testplan.yaml:134]: So still having some issues.  Looks like I was seeing other problems with the 1.5 tests using Nandaa updated rego code.  It looks like the complaint tests were supposed to have an expected return value of true.  But not 100 percent sure my understanding of the test framwork is correct.  I thought that compliant tests should have an expected return of true and non-compiant false respetively.  But this pattern is not followed in tests 4.1 and following.   For example the first 4.1 test is non-complaint and expects false but gets a return true from the rego code.

File: Testing/Functional/Products/TestPlans/teams.testplan.yaml [hunk 3]
```diff
@@ -77,315 +65,367 @@ TestPlan:
...
+      - TestDescription: MS.TEAMS.5.3v1 Non-compliant - BlockedAppList
+        Preconditions:
+          - Command: 'Get-CsTeamsAppPermissionPolicy | % {Set-CsTeamsAppPermissionPolicy -Identity ($_.Identity) -PrivateCatalogAppsType AllowedAppList -PrivateCatalogApps @()}'
+        Postconditions: []
+        ExpectedResult: false
+      - TestDescription: MS.TEAMS.5.3v1 Compliant - AllowedAppList
+        Preconditions:
+          - Command: 'Get-CsTeamsAppPermissionPolicy | % {Set-CsTeamsAppPermissionPolicy -Identity ($_.Identity) -PrivateCatalogAppsType BlockedAppList -PrivateCatalogApps @()}'
+        Postconditions: []
+        ExpectedResult: true
```
Comment (kennethpalmer) [inline Testing/Functional/Products/TestPlans/teams.testplan.yaml:392]:
```suggestion
      - TestDescription: MS.TEAMS.5.3v1 Non-compliant - BlockedAppList
        Preconditions:
          - Command: 'Get-CsTeamsAppPermissionPolicy | % {Set-CsTeamsAppPermissionPolicy -Identity ($_.Identity) -PrivateCatalogAppsType BlockedAppList -PrivateCatalogApps @()}'
        Postconditions: []
        ExpectedResult: false
      - TestDescription: MS.TEAMS.5.3v1 Compliant - AllowedAppList
        Preconditions:
          - Command: 'Get-CsTeamsAppPermissionPolicy | % {Set-CsTeamsAppPermissionPolicy -Identity ($_.Identity) -PrivateCatalogAppsType AllowedAppList -PrivateCatalogApps @()}'
        Postconditions: []
        ExpectedResult: true
```

**Review by kennethpalmer — CHANGES_REQUESTED** [review]

File: Testing/Functional/Products/TestPlans/teams.testplan.yaml (hunk 1 above)
Comment (crutchfield) [inline Testing/Functional/Products/TestPlans/teams.testplan.yaml:56]:
```suggestion
      - TestDescription: MS.TEAMS.1.3v1 Compliant - Dialin not bypass lobby Everyone not auto admitted user
```

Comment (crutchfield) [reply Testing/Functional/Products/TestPlans/teams.testplan.yaml:56]: Corrected

**Review by crutchfield — COMMENTED** [review]
Addressed comment

File: Testing/Functional/Products/TestPlans/teams.testplan.yaml (hunk 2 above)
Comment (crutchfield) [reply Testing/Functional/Products/TestPlans/teams.testplan.yaml]: Removed invalid test cases from 1.5.

File: Testing/Functional/Products/TestPlans/teams.testplan.yaml [hunk 4]
```diff
@@ -77,315 +65,367 @@ TestPlan:
...
+          - Command: 'Get-CsTeamsMeetingPolicy | % {Set-CsTeamsMeetingPolicy -Identity $($_.Identity) -AllowPSTNUsersToBypassLobby $true }'
+        Postconditions: []
+        ExpectedResult: false
+      - TestDescription: MS.TEAMS.1.5v1 Non-compliant - AutoAdmittedUsers = EveryoneInCompanyExcludingGuests
+        Preconditions:
+          - Command: 'Get-CsTeamsMeetingPolicy | % {Set-CsTeamsMeetingPolicy -Identity $($_.Identity) -AutoAdmittedUsers EveryoneInCompanyExcludingGuests }'
+          - Command: 'Get-CsTeamsMeetingPolicy | % {Set-CsTeamsMeetingPolicy -Identity $($_.Identity) -AllowPSTNUsersToBypassLobby $true }'
+        Postconditions: []
+        ExpectedResult: false
+      - TestDescription: MS.TEAMS.1.5v1 Non-compliant - AutoAdmittedUsers = Everyone
```
[the remark this answers is not in the document]
Comment (crutchfield) [reply Testing/Functional/Products/TestPlans/teams.testplan.yaml]: Removed command

**Review by nanda-katikaneni — APPROVED** [review]

**Comment by nanda-katikaneni** [comment]
@crutchfield - one of unit tests is failing - this is the deconflicted policy 1.5. Think the error may be related to the report detail not matching the updated report detail for fail.
>>RuleOutput[0].ReportDetails = "Requirement not met: Dial-in users are enabled to bypass the lobby"
Note that the ReportDetail for this policy failure is changed to: "meeting policy(ies) found that either allow everyone or dial-in users to bypass lobby:"

**Comment by crutchfield** [comment]
@nanda-katikaneni PR is ready to merge.  OPA test failure should be fix by other PR with fix for 1.5v1.

**Pull request merged**